Mtc enhancements 2019v2

landonreed opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JSDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

Addresses some issues with `mtc-enhancements-2019` and adds a couple of other enhancements, including:
- better messages (toast pop ups) when a pattern stop time is changed
- table- and column-level issues for GTFS+
- various validation issue fixes (esp. for blocking feed publication)
- block publication if GTFS+ validation issues present
- allow deletion of GTFS+ edits/changes

evansiroky commented on lib/common/components/Loading.js:
```diff
@@ -5,20 +5,27 @@ import React, {Component} from 'react'
+        className={`fa-spin ${small ? '' : 'fa-5x'}`}
+        type='refresh' />
+    )
+    return small
```
What does `small` do here? Could use a comment.

landonreed replied:
The `small` prop is being used both to make the icon smaller by excluding the `fa-5x` class and is also being used to return the icon inline (e.g., directly next to some other text) rather than on its own row. Do you think this should be a new prop called `inline` that is specified separately from `small`?

evansiroky replied:
Yeah, maybe call it inline. Whatever the case, please do add a comment.

evansiroky commented on lib/common/containers/ActiveSidebarNavItem.js:
```diff
@@ -21,8 +21,8 @@ const mapStateToProps = (state: AppState, ownProps: Props) => {
 const mapDispatchToProps = {}
-
-var ActiveSidebarNavItem = connect(
+// $FlowFixMe https://github.com/flow-typed/flow-typed/issues/2628
```
Trying to resolve this flow error is hella annoying. Guess I have to let it slide.

evansiroky commented on lib/editor/actions/tripPattern.js:
```diff
@@ -2,9 +2,11 @@
 
 import {resetActiveGtfsEntity, savedGtfsEntity, updateActiveGtfsEntity, updateEditSetting} from './active'
 import {createVoidPayloadAction, secureFetch} from '../../common/actions'
+// import {showToast} from '../../common/actions/status'
```
Please remove

evansiroky commented on lib/gtfs/components/gtfs-search.js:
```diff
@@ -188,6 +188,7 @@ const mapDispatchToProps = {}
 
 const mapStateToProps = (state: AppState, ownProps) => ({user: state.user})
 
+// $FlowFixMe https://github.com/flow-typed/flow-typed/issues/2628
```
Also quite annoying

evansiroky commented on lib/gtfsplus/components/GtfsPlusTable.js:
```diff
@@ -8,15 +8,15 @@ import GtfsPlusField from './GtfsPlusField'
   showHelpClicked: (string, ?string) => void,
   table: GtfsPlusTableType,
-  tableValidation: Array<any>
+  tableValidation: Array<GtfsPlusValidationIssue>
```
Nice!

evansiroky commented on lib/gtfsplus/components/GtfsPlusVersionSummary.js:
```diff
@@ -43,35 +45,44 @@ export default class GtfsPlusVersionSummary extends Component<Props, State> {
-    if (!this.gtfsPlusEdited()) return <i>GTFS+ data for this feed version has not been edited.</i>
-    return <b>GTFS+ Data updated {moment(timestamp).format('MMM. DD, YYYY, h:MMa')}</b>
+    if (!this._gtfsPlusIsEdited()) return <i>GTFS+ data for this feed version has not been edited.</i>
+    const ts = moment(timestamp)
```
Please use a more descriptive variable name

evansiroky commented on lib/manager/components/version/FeedVersionNavigator.js:
```diff
@@ -315,6 +310,7 @@ export default class FeedVersionNavigator extends Component<Props, State> {
               publishFeedVersion={publishFeedVersion}
               renameFeedVersion={renameFeedVersion}
               user={user}
+              validateGtfsPlusFeed={this.props.validateGtfsPlusFeed}
```
Holy cow that's a lot of props. It might be nice to connect a few of the child components so all this data isn't passed through. Something for later.

landonreed replied:
I'll create a ticket for this to revisit: #455

evansiroky commented on lib/manager/components/version/FeedVersionViewer.js:
```diff
@@ -43,6 +46,7 @@ export type Props = {
   publishFeedVersion: typeof versionsActions.publishFeedVersion,
   renameFeedVersion: typeof versionsActions.renameFeedVersion,
   user: ManagerUserState,
+  validateGtfsPlusFeed: typeof plusActions.validateGtfsPlusFeed,
```
Again, way more props than are needed.

evansiroky commented on lib/manager/components/version/FeedVersionViewer.js:
```diff
@@ -51,13 +55,32 @@ export type Props = {
       newNotePosted,
       notesRequested,
       version,
+      versions,
```
Any of the components in this `render` function that get passed `{...this.props}` really ought to be connected components.

evansiroky commented on lib/manager/util/version.js:
```diff
@@ -35,3 +40,21 @@ export function versionHasExpired (version: FeedVersion): boolean {
+  'REFERENTIAL_INTEGRITY'
+]
+
+export const versionsComparator = (a: FeedVersion, b: FeedVersion) => {
```
Could have a more specific name indicating it's comparing the updated attribute.

evansiroky commented on lib/manager/components/version/FeedVersionDetails.js:
```diff
@@ -85,7 +105,17 @@ export default class FeedVersionDetails extends Component<Props> {
+                  isPublished ||
+                  processing ||
+                  !userCanManageFeed
+                }
```
It'd be nice for there to be a tooltip or some explainer on why this button is disabled since there are a lot of reasons why.

evansiroky replied:
I see there is something below if there are blocking issues, but there are more reasons up here that cause the button to be disabled.

landonreed replied:
Many of the reasons listed here (which are not shown in the message below) are more so related to the fact that things have not fully loaded yet.

landonreed commented:
Thanks, @evansiroky. I've addressed all of the comments and created a new issue (#455) to account for the needed refactor, which I'd like to address outside of this PR.

landonreed commented:
:tada: This PR is included in version 4.1.0 :tada:

The release is available on [GitHub release](https://github.com/ibi-group/datatools-ui/releases/tag/v4.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
